### ISSUE TITLE
Enable formatting in product description

### ISSUE THREAD

**Issue by myriamboure** [openfoodfoundation/openfoodnetwork#1020]
Continuing from https://github.com/openfoodfoundation/openfoodnetwork/issues/953

A request has come from our users to be able to format product description (basic: bold, list, go to the next line, etc.)
Today it's not very user friendly...

![image](https://cloud.githubusercontent.com/assets/8135376/15666602/ece57f60-2711-11e6-8b5b-89d84f5aa890.png)

**Comment by myriamboure** [comment]
Why not using styling with markdown for example? Like it's done in github :-)

**Comment by myriamboure** [comment]
Any solution on that @sstead ? Did you have any request from users to be able to format the product description?

**Comment by sstead** [comment]
Hi @myriamboure, I ran this idea past Rob and Rohan and they said there are a couple of different options. Wizzywig would be the easiest option in the short term (this is what we have for About Us profile description). However this needs to be thought out, because it will mean saving data in HTML, from which point it's hard to go back. Markdown is also appealing, but again needs thought. So essentially, yes there are options for doing this, it's not a huge task but it will need some consideration, and some resources to make happen. We haven't had requests from users for this feature in Australia.

**Comment by OliverUK** [comment]
Thanks for pointing me here @myriamboure! I've added the UK label.

**Comment by lin-d-hop** [comment]
@OliverUK Just wanted to check why you've added this to UK Current? Is it an urgent priority or something Stroudco in particular would like to see?

**Comment by OliverUK** [comment]
@lin-d-hop yes. It's not critical but certainly seems the lack of any formatting results in a rather incomplete package for hubs and users. Here you want to be able to sell your items, write about it, possible add links or a list. Currently it's really only good enough for one or two sentences. It doesn't even do paragraphs.

**Comment by lin-d-hop** [comment]
Ok. I agree it can make a big difference, particularly for producers. Let's get it explored.

@stveep (or @lebreeze)
Would it be possible to have a few options put on the table here? Would you be able to explore options for formatting the product descriptions including enabling HTML and markdown with pros and cons of each and a recommendation of the best course of action. 
It would be good to have flexibility to upgrade or change later if requirements change so please bear this in mind. 
Usability is very important. From a UX perspective we want to make it very easy for users to create attractive product descriptions. 

We interact with a number of examples ourselves, within github and discourse... so we have an idea of what works well :-)

**Comment by lin-d-hop** [comment]
Tagging @levent as I tagged incorrectly above.

**Comment by Matt-Yorkley** [comment]
Is there much advantage to using Markup over HTML? A version of the "textAngular" WYSIWYG package and it's corresponding css files are already used in some of the admin sections:

https://github.com/openfoodfoundation/openfoodnetwork/blob/0760d4cc1f26fd06f60b420293e3671f205d0a07/app/views/admin/enterprises/form/_about_us.html.haml

Should be quite easy to add it to the product descriptions, and re-using it would be good for UI consistency. I could have a go at it this week.

**Comment by NickWeir63** [comment]
Hi Matt
Great to see you back involved in OFN dev stuff.  Thanks
Would you be able to join our dev update hangout tomorrow morning?  It is a fortnightly update where we all check in with input from Danni (our contact with the Aus dev team).  We all do it on a voluntary basis but it is a good way of getting support and finding out what else is going on.
Here is [the agenda for the hangout ](https://docs.google.com/spreadsheets/d/1p_BDNgLPyCzOsZO7P_4nFkpRA__ARz6v9Up7oqEIKdw/edit#gid=1768957062)
We open a `[hangout on Slack here](https://openfoodnetwork.slack.com/messages/uk-au-chats/team/)` just before the meeting which will start at 8.30am tomorrow (Weds)
Hope you can join us
Thanks
Nick

**Comment by Matt-Yorkley** [comment]
I've made an implementation for this but the textAngular javascript isn't instantiating properly on the edit product page. It's loaded properly into the admin.products module and works perfectly on the create product page. I can't tell if there's a conflict with other scripts on the page or not, and there's no errors in my javascript debugger...

https://github.com/openfoodfoundation/openfoodnetwork/compare/master...Matt-Yorkley:product_descriptions

Any ideas?

**Comment by Matt-Yorkley** [comment]
The fact that it works on the create product page but not the edit product page is confusing. textAngular is supposed to find any elements labelled <text-angular> or <div text-angular> in the page and insert the additional WYSIWYG elements. I've checked the textAngular.min.js file is definitely loading on the edit products page. I've tried making dummy textAngular instances on the page that don't even use the product data model and they don't work either, so I'm thinking some sort of conflict is stopping it running properly?

**Comment by daniellemoorhead** [comment]
@rob @mkllnk thoughts?

**Comment by Matt-Yorkley** [comment]
Tried temporarily adding some javascript to the top of textAngular.min.js that removes the text-angular element from the DOM on page load, confirmed both that the script is running on the page and the element is present when the script loads (in case an issue with Deface was causing the element to load after the script)...

**Comment by mkllnk** [comment]
That sounds strange. Can you see any difference in usage compared to the other pages that use text-angular? Do you need an ng-model defined?

**Comment by Matt-Yorkley** [comment]
On the add product page it works whilst ng-model or any of the other attributes are stripped out completely. I've tried every combination I can think of but it never loads on the edit products page. Baffling...

**Comment by stveep** [comment]
Hi @Matt-Yorkley I managed to fix this by adding a separate script to bootstrap the angular app - this code was at the bottom of the deface for the new product page where the text box was working. I think these older Spree views don't include any angular by default. See my PR on your branch.

There might be a more general way to do this, but most of the admin pages have their own angular module so I guess you at least need a different command for each admin page...

**Comment by Matt-Yorkley** [comment]
Thanks @stveep, I was lost on that one. So it looks like angular isn't instantiated at all on that page? Maybe just adding ng-app to the form would be more elegant than the javascript block at the bottom?

**Comment by stveep** [comment]
I have found myself having to do it this way before - I think when the page has to wait for some data to load before the angular app can be started (hence the `.ready()` in the script that I added) - I guess here all the data is there in the rails view so maybe we can do it with the ng-app directive as you say. I did try that first when I was messing around with it but I might have had the module name wrong at that stage so maybe worth another go.

@oeoeaio or @mkllnk do you have any opinions/guidance on when to use the different bootstrapping methods in the admin pages?

**Comment by Matt-Yorkley** [comment]
I've done it with ng-app now and it's working fine. I've finished this implementation of the the formatted product descriptions now, but it'd be good to get a bit of code review and feedback for it. I had to make a few executive decisions that might not be optimal. 

For instance: on the frontend I've added a bit to the product serializer to optionally return html, but left the standard description attribute intact with tags stripped in case it's wanted/used somewhere else that doesn't want html:

![screenshot from 2016-11-27 14-13-13](https://cloud.githubusercontent.com/assets/9029026/20648979/a77ab41c-b4ac-11e6-9ffc-c3a46ffc24c4.png)

**Comment by lin-d-hop** [comment]
Thanks for this @Matt-Yorkley and also for the help @stveep 

@stveep Would you be happy to review this one? Or might it be better to get a 3rd set of eyes on the issue as it sounds like it was kinda quirky?

**Comment by OliverUK** [comment]
@lin-d-hop @Matt-Yorkley @stveep 

I've logged into staging using super admin (mine didn't work) on a Safari browser on Mac. I created an OC and added suppliers before editing one item by clicking the edit button in BEP. 

![screen shot 2016-12-08 at 12 09 16](https://cloud.githubusercontent.com/assets/17590921/21009666/bb858bf8-bd3f-11e6-9d50-7ed8486f3709.png)

I was able to edit the description and save.
Here is how it came out in the shop front before and after.

![screen shot 2016-12-08 at 12 08 27](https://cloud.githubusercontent.com/assets/17590921/21009690/d7a8cd18-bd3f-11e6-8938-621284486e19.png)

![screen shot 2016-12-08 at 12 10 04](https://cloud.githubusercontent.com/assets/17590921/21009695/dbd4e8f4-bd3f-11e6-9654-796b014a5c61.png)

The BOLD text works fine but a line break creates a paragraph. 

I then created a new product and gave it the following description.
![screen shot 2016-12-08 at 12 18 58](https://cloud.githubusercontent.com/assets/17590921/21009849/befd6598-bd40-11e6-87d3-70be84406497.png)

Here is the result

![screen shot 2016-12-08 at 12 22 41](https://cloud.githubusercontent.com/assets/17590921/21009916/18d5583c-bd41-11e6-8809-447c18b06201.png)

**Comment by Matt-Yorkley** [comment]
I have a separate pull request that fixes the discrepancies between how line breaks appear in the WYSIWYG and on the backend. I think it applies to a couple of other things as well, like enterprise descriptions. It's not included yet, but it's been addressed.

**Comment by lin-d-hop** [comment]
When u submit it can u ensure the branch fits the pattern uk/* 
Will this need a separate code review?
Cheers matt!

**Comment by Matt-Yorkley** [comment]
It's in #1238 from November. Stveep's had a look already. It might need some new specs though.

**Comment by lin-d-hop** [comment]
Ok so this issue is being blocked by #1238 which needs cleaning up and testing.

**Comment by lin-d-hop** [comment]
The outstanding issues here have been fixed in a separate PR #1327 
Adding this to v1.7.
Will get #1327 in there too for the next release.

**Comment by lin-d-hop** [comment]
@sstead Can you confirm just confirm you are happy with the test results? Thanks!

**Comment by lin-d-hop** [comment]
Back to @Matt-Yorkley after @sstead comments in the PR.

**Comment by daniellemoorhead** [comment]
Sorry guys, @sstead had a look at this but there were "problems" (end quote). @oeoeaio is looking at it today but it won't make it into the release. Next year :)

**Comment by daniellemoorhead** [comment]
Looks like the only PR remaining for this issue is #1239 and there are conflicts with it. Thus am moving this issue to 1.8.9 release.

**Comment by daniellemoorhead** [comment]
PR #1239 is in code review and testing with Rob and Sally. All going to plan will be part of 1.8.9 release
